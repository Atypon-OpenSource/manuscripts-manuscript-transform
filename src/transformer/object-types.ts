/*!
 * Â© 2019 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  Figure,
  Manuscript,
  manuscriptIDTypes,
  Model,
  ObjectTypes,
  Table,
} from '@manuscripts/manuscripts-json-schema'

import { ManuscriptModel, UserProfileWithAvatar } from './models'

export enum ExtraObjectTypes {
  PlaceholderElement = 'MPPlaceholderElement',
}

export const elementObjects = [
  ObjectTypes.BibliographyElement,
  ObjectTypes.EquationElement,
  ObjectTypes.FigureElement,
  ObjectTypes.FootnotesElement,
  ObjectTypes.ListElement,
  ObjectTypes.ListingElement,
  ObjectTypes.ParagraphElement,
  ObjectTypes.TableElement,
  ObjectTypes.TOCElement,
]

export const manuscriptObjects = [
  ObjectTypes.Affiliation,
  ObjectTypes.Citation,
  ObjectTypes.CommentAnnotation,
  ObjectTypes.Contributor,
  ObjectTypes.Footnote,
  ObjectTypes.InlineMathFragment,
  ObjectTypes.Section,
].concat(elementObjects) // TODO: remove elementObjects if they don't need `manuscriptID`

export const isManuscriptModel = (model: Model): model is ManuscriptModel => {
  // TODO: check all required fields
  if (!model.objectType) {
    throw new Error('Model must have objectType')
  }

  return manuscriptIDTypes.has(model.objectType)
}

export const hasObjectType = <T extends Model>(objectType: string) => (
  model: Model
): model is T => model.objectType === objectType

export const isFigure = hasObjectType<Figure>(ObjectTypes.Figure)
export const isManuscript = hasObjectType<Manuscript>(ObjectTypes.Manuscript)
export const isTable = hasObjectType<Table>(ObjectTypes.Table)
export const isUserProfile = hasObjectType<UserProfileWithAvatar>(
  ObjectTypes.UserProfile
)
