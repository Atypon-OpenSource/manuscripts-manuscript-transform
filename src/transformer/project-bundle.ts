/*!
 * Â© 2019 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  Journal,
  Manuscript,
  Model,
  ObjectTypes,
  Title,
} from '@manuscripts/json-schema'

import { Decoder } from './decode'
import { ManuscriptModel } from './models'
import { hasObjectType } from './object-types'

export interface ProjectBundle {
  version: string
  data: Model[]
}

export const parseProjectBundle = (
  projectBundle: ProjectBundle,
  manuscriptID?: string,
) => {
  const manuscriptData = manuscriptID
    ? (projectBundle.data as ManuscriptModel[]).filter(
        (doc) => !doc.manuscriptID || doc.manuscriptID === manuscriptID
      )
    : projectBundle.data

  const modelMap: Map<string, Model> = new Map()

  for (const component of manuscriptData) {
    modelMap.set(component._id, component)
  }

  const decoder = new Decoder(modelMap)

  const doc = decoder.createArticleNode()

  return { doc, modelMap }
}

const isManuscript = hasObjectType<Manuscript>(ObjectTypes.Manuscript)

export const findManuscript = (modelMap: Map<string, Model>): Manuscript => {
  for (const model of modelMap.values()) {
    if (isManuscript(model)) {
      return model
    }
  }

  throw new Error('No manuscript found')
}

const isJournal = hasObjectType<Journal>(ObjectTypes.Journal)

export const findJournal = (modelMap: Map<string, Model>) => {
  for (const model of modelMap.values()) {
    if (isJournal(model)) {
      return model
    }
  }

  return null
}

const isManuscriptModel = (model: Model): model is ManuscriptModel =>
  'manuscriptID' in model

export const findManuscriptModelByType = <T extends ManuscriptModel>(
  modelMap: Map<string, Model>,
  manuscript: Manuscript,
  objectType: ObjectTypes
): T | undefined => {
  for (const model of modelMap.values()) {
    if (
      model.objectType === objectType &&
      isManuscriptModel(model) &&
      manuscript._id === model.manuscriptID
    ) {
      return model as T
    }
  }
}

export const findManuscriptById = (
  modelMap: Map<string, Model>,
  manuscriptID: string
): Manuscript => {
  const manuscript: Model | undefined = modelMap.get(manuscriptID)

  if (manuscript && isManuscript(manuscript)) {
    return manuscript as Manuscript
  }

  throw new Error(
    `There is no manuscript found for the following _id (${manuscriptID})`
  )
}

const isTitle = hasObjectType<Title>(ObjectTypes.Journal)

export const findTitle = (modelMap: Map<string, Model>): Title => {
  for (const model of modelMap.values()) {
    if (isTitle(model)) {
      return model
    }
  }

  throw new Error('No Title found')
}
