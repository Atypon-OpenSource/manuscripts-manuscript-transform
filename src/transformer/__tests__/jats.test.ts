/*!
 * © 2019 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import projectDumpWithCitations from '@manuscripts/examples/data/project-dump-2.json'
import projectDump from '@manuscripts/examples/data/project-dump.json'
import {
  ObjectTypes,
  ParagraphElement,
  Section,
} from '@manuscripts/manuscripts-json-schema'
import { parseXml } from 'libxmljs2'
import { JATSTransformer } from '../jats'
import { isFigure } from '../object-types'
import { parseProjectBundle, ProjectBundle } from '../project-bundle'
import { submissions } from './__helpers__/submissions'

const input = projectDump as ProjectBundle
const inputWithCitations = projectDumpWithCitations as ProjectBundle

const cloneProjectBundle = (input: ProjectBundle): ProjectBundle =>
  JSON.parse(JSON.stringify(input))

const parseXMLWithDTD = (data: string) =>
  parseXml(data, {
    dtdload: true,
    dtdvalid: true,
    nonet: true,
  })

describe('jats', () => {
  test('export latest version', () => {
    const projectBundle = cloneProjectBundle(input)

    const { doc, modelMap } = parseProjectBundle(projectBundle)

    const transformer = new JATSTransformer()
    const result = transformer.serializeToJATS(doc.content, modelMap)

    expect(result).toMatchSnapshot('jats-export')
  })

  test('export v1.1', () => {
    const projectBundle = cloneProjectBundle(input)

    const { doc, modelMap } = parseProjectBundle(projectBundle)

    const transformer = new JATSTransformer()
    const result = transformer.serializeToJATS(doc.content, modelMap, '1.1')

    expect(result).toMatchSnapshot('jats-export-1.1')
  })

  test('export unknown version', () => {
    const projectBundle = cloneProjectBundle(input)

    const { doc, modelMap } = parseProjectBundle(projectBundle)

    expect(() => {
      // @ts-ignore (deliberately invalid)
      serializeToJATS(doc.content, modelMap, '1.0')
    }).toThrow()
  })

  test('move abstract to front by section category', () => {
    const projectBundle = cloneProjectBundle(input)

    const model: Section = {
      _id: 'MPSection:123',
      objectType: 'MPSection',
      createdAt: 0,
      updatedAt: 0,
      category: 'MPSectionCategory:abstract',
      title: 'Foo',
      manuscriptID: 'MPManuscript:1',
      containerID: 'MPProject:1',
      path: ['MPSection:123'],
      sessionID: 'foo',
      priority: 0,
    }

    projectBundle.data.push(model)

    const { doc, modelMap } = parseProjectBundle(projectBundle)

    const transformer = new JATSTransformer()
    const xml = transformer.serializeToJATS(doc.content, modelMap)

    const resultDoc = parseXMLWithDTD(xml)

    const result = resultDoc.get('/article/front/article-meta/abstract')

    expect(result).not.toBeNull()
  })

  test('move abstract to front by title', () => {
    const projectBundle = cloneProjectBundle(input)

    const model: Section = {
      _id: 'MPSection:123',
      objectType: 'MPSection',
      createdAt: 0,
      updatedAt: 0,
      category: '',
      title: 'Abstract',
      manuscriptID: 'MPManuscript:1',
      containerID: 'MPProject:1',
      path: ['MPSection:123'],
      sessionID: 'foo',
      priority: 3,
    }

    projectBundle.data.push(model)

    const { doc, modelMap } = parseProjectBundle(projectBundle)

    const transformer = new JATSTransformer()
    const xml = transformer.serializeToJATS(doc.content, modelMap)

    const resultDoc = parseXMLWithDTD(xml)

    const result = resultDoc.get('/article/front/article-meta/abstract')

    expect(result).not.toBeNull()
  })

  test('handle ID', () => {
    const projectBundle = cloneProjectBundle(input)

    const { doc, modelMap } = parseProjectBundle(projectBundle)

    const transformer = new JATSTransformer()
    const result = transformer.serializeToJATS(
      doc.content,
      modelMap,
      '1.2',
      undefined,
      '123'
    )

    expect(result).toMatchSnapshot('jats-export-id')
  })

  test('handle DOI', () => {
    const projectBundle = cloneProjectBundle(input)

    const { doc, modelMap } = parseProjectBundle(projectBundle)

    const transformer = new JATSTransformer()
    const result = transformer.serializeToJATS(
      doc.content,
      modelMap,
      '1.2',
      '10.0000/123'
    )

    expect(result).toMatchSnapshot('jats-export-doi')
  })

  test('add journal ID', () => {
    const projectBundle = cloneProjectBundle(input)

    const { doc, modelMap } = parseProjectBundle(projectBundle)

    for (const submission of submissions) {
      modelMap.set(submission._id, submission)
    }

    const transformer = new JATSTransformer()
    const xml = transformer.serializeToJATS(
      doc.content,
      modelMap,
      '1.2',
      '10.0000/123',
      '123'
    )

    expect(xml).toMatchSnapshot('jats-export-submitted')

    const output = parseXMLWithDTD(xml)

    expect(output.get('//journal-id')!.text()).toBe('bar')
    expect(output.get('//journal-title')!.text()).toBe('Bar')
    expect(output.get('//issn')!.text()).toBe('2222-2222')
  })

  test('DTD validation', () => {
    const projectBundle = cloneProjectBundle(input)

    const { doc, modelMap } = parseProjectBundle(projectBundle)

    const transformer = new JATSTransformer()
    const xml = transformer.serializeToJATS(doc.content, modelMap)

    const { errors } = parseXMLWithDTD(xml)

    expect(errors).toHaveLength(0)
  })

  test('DTD validation: article with title markup and citations', () => {
    const projectBundle = cloneProjectBundle(inputWithCitations)

    const { doc, modelMap } = parseProjectBundle(projectBundle)

    const transformer = new JATSTransformer()
    const xml = transformer.serializeToJATS(doc.content, modelMap)

    const { errors } = parseXMLWithDTD(xml)

    expect(errors).toHaveLength(0)
  })

  test('Export link', () => {
    const projectBundle = cloneProjectBundle(input)

    const id = 'MPParagraphElement:150780D7-CFED-4529-9398-77B5C7625044'

    projectBundle.data = projectBundle.data.map(model => {
      if (model._id === id) {
        const paragraphElement = model as ParagraphElement

        paragraphElement.contents = paragraphElement.contents.replace(
          /The first section/,
          'The <a href="https://example.com">first</a> section'
        )
      }

      return model
    })

    const { doc, modelMap } = parseProjectBundle(projectBundle)

    const transformer = new JATSTransformer()
    const xml = transformer.serializeToJATS(doc.content, modelMap)

    const { errors } = parseXMLWithDTD(xml)

    expect(errors).toHaveLength(0)

    const output = parseXMLWithDTD(xml)

    const link = output.get('//ext-link[@ext-link-type="uri"]')

    expect(link).not.toBeNull()
    expect(link!.text()).toBe('first')

    const attrs: { [key: string]: string } = {}

    for (const attr of link!.attrs()) {
      attrs[attr.name()] = attr.value()
    }

    expect(attrs.href).toBe('https://example.com')
  })

  test('Export with missing bibliography element', () => {
    const projectBundle = cloneProjectBundle(input)

    const id = 'MPSection:E07B0D52-9642-4D58-E577-26F8804E3DEE'

    projectBundle.data = projectBundle.data.filter(
      model =>
        model.objectType !== ObjectTypes.BibliographyElement && model._id !== id
    )

    const { doc, modelMap } = parseProjectBundle(projectBundle)

    const transformer = new JATSTransformer()
    const xml = transformer.serializeToJATS(doc.content, modelMap)

    const { errors } = parseXMLWithDTD(xml)

    expect(errors).toHaveLength(0)

    const output = parseXMLWithDTD(xml)

    const refs = output.find('//ref-list/ref')

    expect(refs).toHaveLength(1)
  })

  test('Markup in citations', () => {
    const projectBundle = cloneProjectBundle(inputWithCitations)

    const { doc, modelMap } = parseProjectBundle(projectBundle)

    const transformer = new JATSTransformer()
    const xml = transformer.serializeToJATS(doc.content, modelMap)

    const output = parseXMLWithDTD(xml)

    const refs = output.find('//xref[@ref-type="bibr"]')

    expect(refs).toHaveLength(2)

    expect(refs[0].child(0)!.type()).toBe('text')
    expect(refs[0].text()).toBe('1,2')
    expect(refs[1].child(0)!.type()).toBe('text')
    expect(refs[1].text()).toBe('3–5')
  })

  test('Export with empty figure', () => {
    const projectBundle = cloneProjectBundle(input)

    for (const model of projectBundle.data) {
      if (isFigure(model)) {
        delete model._id
        break
      }
    }

    const { doc, modelMap } = parseProjectBundle(projectBundle)

    const transformer = new JATSTransformer()
    const xml = transformer.serializeToJATS(doc.content, modelMap)

    const { errors } = parseXMLWithDTD(xml)
    expect(errors).toHaveLength(0)

    const output = parseXMLWithDTD(xml)

    const figures = output.find('//fig')
    expect(figures).toHaveLength(3)

    const figureGroups = output.find('//fig-group')
    expect(figureGroups).toHaveLength(0)
  })

  test('Only export front matter', () => {
    const projectBundle = cloneProjectBundle(input)

    const { doc, modelMap } = parseProjectBundle(projectBundle)

    const transformer = new JATSTransformer()
    const xml = transformer.serializeToJATS(
      doc.content,
      modelMap,
      '1.2',
      '10.1234/5678',
      '4567',
      true
    )

    const { errors } = parseXMLWithDTD(xml)
    expect(errors).toHaveLength(0)

    const output = parseXMLWithDTD(xml)

    const front = output.find('//front')
    expect(front).toHaveLength(1)

    const doi = output.find('//article-id[@pub-id-type="doi"]')
    expect(doi).toHaveLength(1)

    const body = output.find('//body')
    expect(body).toHaveLength(0)

    const back = output.find('//back')
    expect(back).toHaveLength(0)
  })
})
