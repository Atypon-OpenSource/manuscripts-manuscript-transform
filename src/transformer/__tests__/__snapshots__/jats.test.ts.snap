// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jats export latest version: jats-export 1`] = `
"<!DOCTYPE article PUBLIC \\"-//NLM//DTD JATS (Z39.96) Journal Archiving and Interchange DTD with OASIS Tables with MathML3 v1.2 20190208//EN\\" \\"http://jats.nlm.nih.gov/archiving/1.2/JATS-archive-oasis-article1-mathml3.dtd\\"><article xmlns:xlink=\\"http://www.w3.org/1999/xlink\\"><front><journal-meta><journal-id journal-id-type=\\"publisher-id\\"/><journal-title-group><journal-title/></journal-title-group><issn pub-type=\\"epub\\"/></journal-meta><article-meta><title-group><article-title>Example Manuscript</article-title></title-group><contrib-group content-type=\\"authors\\"><contrib contrib-type=\\"author\\" id=\\"MPContributor_585DB23A-8778-4AFF-986F-CFF7B733CDE6\\"><name><surname>Author</surname><given-names>Example</given-names></name><email>user@example.com</email><xref ref-type=\\"aff\\" rid=\\"MPAffiliation_DED56C09-42E2-4AB4-BA15-9260389E2B08\\"/></contrib><contrib contrib-type=\\"author\\" id=\\"MPContributor_8AD5903E-0C9C-4F46-87AE-124142CE12C2\\" corresp=\\"yes\\"><name><surname>Author</surname><given-names>Corresponding</given-names></name><email>author@example.com</email></contrib></contrib-group><aff id=\\"MPAffiliation_DED56C09-42E2-4AB4-BA15-9260389E2B08\\"><institution>University of Examples</institution><country>United Kingdom</country></aff></article-meta></front><body><sec id=\\"MPSection_DFCBE345-A12D-451E-8320-723F25D2F472\\"><title>Table of Contents</title><div id=\\"MPTOCElement_657C54DB-78B5-4206-E322-170F68263A01\\"/></sec><sec id=\\"MPSection_77786D47-6060-4FBC-BC13-5FA754968D6A\\"><title>First section</title><p id=\\"MPParagraphElement_150780D7-CFED-4529-9398-77B5C7625044\\">The first section.</p><p id=\\"MPParagraphElement_60C13548-23C0-4348-FE98-0510E0B61B18\\">The text in this sentence is <bold>bold</bold>, <italic>italic</italic>, or <underline>underlined</underline>.</p><p id=\\"MPParagraphElement_A89DC62A-90ED-4283-AB6D-06CA2519D801\\"><bold><italic><underline>The text in this sentence is bold, italic and underlined.</underline></italic></bold></p><p id=\\"MPParagraphElement_05A0ED43-8928-4C69-A17C-0A98795001CD\\">The text in this sentence is <sup>superscript</sup> and <sub>subscript</sub>.</p></sec><sec id=\\"MPSection_F8EE617B-484B-4CF2-F1CA-014F5307D5CF\\"><title>Section Two</title><p id=\\"MPParagraphElement_4DF93A63-D44D-41CF-AF54-23712B8703FA\\">The heading of this section is hidden.</p><list list-type=\\"ordered\\"><list-item><p>A numbered list.</p></list-item></list><list list-type=\\"bullet\\"><list-item><p>A bullet list.</p></list-item></list><list list-type=\\"ordered\\"><list-item><p>A numbered list with a nested list.</p><list list-type=\\"ordered\\"><list-item><p>This item is nested.</p></list-item></list></list-item></list><fig-group id=\\"MPFigureElement_A5D68C57-B5BB-4D10-E0C3-ECED717A2AA7\\"><caption><p>A figure with a caption</p></caption><fig id=\\"MPFigure_D673DF08-D75E-4061-8EC1-9611EAB302F0\\"><caption><p/></caption><graphic xlink:href=\\"Data/MPFigure_D673DF08-D75E-4061-8EC1-9611EAB302F0.png\\" mimetype=\\"image\\" mime-subtype=\\"png\\"/></fig></fig-group><fig-group id=\\"MPFigureElement_B32BDEC8-E816-4946-C0BB-0404AF2F0C29\\"><caption><p>A figure with hidden caption</p></caption><fig id=\\"MPFigure_AD65D628-A904-4DC4-A026-F8F4C08F6557\\"><caption><p/></caption><graphic xlink:href=\\"Data/MPFigure_AD65D628-A904-4DC4-A026-F8F4C08F6557.png\\" mimetype=\\"image\\" mime-subtype=\\"png\\"/></fig></fig-group><table-wrap id=\\"MPTableElement_82BD8BAF-D136-4536-AD9C-92C30B9D7315\\"><caption><p>A table</p></caption><table id=\\"MPTable_509A944E-994C-4867-AE55-593628C5DBD5\\"><tr><td>Table Header</td><td>Table Header</td></tr><tr><td>Table Data</td><td>Table Data</td></tr><tr><td>Table Data</td><td>Table Data</td></tr><tr><td>Table Footer</td><td>Table Footer</td></tr></table></table-wrap><table-wrap id=\\"MPTableElement_E0D445B4-18C4-43D3-97B3-1076A7BB5CF4\\"><caption><p>A table with hidden header</p></caption><table id=\\"MPTable_7E83812B-A759-4D38-C1C5-F6507D2FC2A9\\"><tr><td>Table Header</td><td/></tr><tr><td/><td/></tr><tr><td/><td/></tr><tr><td/><td/></tr></table></table-wrap><table-wrap id=\\"MPTableElement_1DF69096-DBA3-47B0-B8C2-E06491E4C655\\"><caption><p>A table with hidden footer</p></caption><table id=\\"MPTable_2A2413E2-71F5-4B6C-F513-7B44748E49A8\\"><tr><td/><td/></tr><tr><td/><td/></tr><tr><td/><td/></tr><tr><td/><td/></tr></table></table-wrap><table-wrap id=\\"MPTableElement_89CF51B2-A20C-4AB9-CBD9-F47DAF167DAA\\"><caption><p>A table with hidden caption</p></caption><table id=\\"MPTable_6C045963-404A-4E1E-FF81-B40E9772A6A6\\"><tr><td/><td/></tr><tr><td/><td/></tr><tr><td/><td/></tr><tr><td/><td/></tr></table></table-wrap><fig id=\\"MPEquationElement_3896A81A-D43B-483F-B990-F64518A940D1\\"><caption><p/></caption><disp-formula><tex-math>2x3=6</tex-math></disp-formula></fig><fig id=\\"MPListingElement_160785A9-4BC0-4C55-C4DE-AF5F4D858DCE\\"><caption><p/></caption><code id=\\"MPListing_4D688DCF-29CD-40CC-93BD-2B13CD883257\\" language=\\"javascript\\">// a JavaScript code block

const x = 'y'</code></fig><p id=\\"MPParagraphElement_345B8B57-2E5E-45ED-E587-4EAAC2BD63E9\\">An inline equation: <inline-formula><tex-math>3+5=8</tex-math></inline-formula></p><p id=\\"MPParagraphElement_6B6DBF43-F7B2-404C-89DD-0EC1E1442B27\\">A cross-reference to a figure: <xref ref-type=\\"fig\\" rid=\\"MPAuxiliaryObjectReference_82E2F9F9-797F-4397-9084-D98B72C12903\\">Figure 1</xref></p><p id=\\"MPParagraphElement_41B99DAB-F86C-477F-D36C-2B8590B5DAD0\\">A cross-reference to a table: <xref ref-type=\\"fig\\" rid=\\"MPAuxiliaryObjectReference_88AF1DCA-B232-4791-8EAA-F38E0025A3C6\\">Table 1</xref></p><p id=\\"MPParagraphElement_C8A173A4-DF32-4242-E800-28A1AE5B48B8\\">A cross-reference to an equation: <xref ref-type=\\"fig\\" rid=\\"MPAuxiliaryObjectReference_461D1AFF-1871-4594-8E07-963D3476B7F5\\">Equation 1</xref></p><p id=\\"MPParagraphElement_CB7347E5-8502-4667-82B9-51104BB1FEDB\\"><xref ref-type=\\"bibr\\" rid=\\"MPBibliographyItem_8C394C86-F7B0-48CE-D5BC-E7A10FCE7FA5\\">A hyperlink</xref></p><p id=\\"MPParagraphElement_3708E8C7-441C-4502-8BFC-D3A2059CEF8B\\">A footnote:<xref ref-type=\\"fn\\" rid=\\"MPFootnote_B1589279-0F3B-472B-BF11-3F59D6B65791\\">1</xref></p><fn-group id=\\"MPFootnotesElement_4CD7CE08-0880-46BD-FF83-52E32AEFAC3E\\"/></sec></body><back><ref-list id=\\"MPSection_E07B0D52-9642-4D58-E577-26F8804E3DEE\\"><title>Bibliography</title><ref id=\\"MPBibliographyItem_8C394C86-F7B0-48CE-D5BC-E7A10FCE7FA5\\"><element-citation publication-type=\\"journal\\"><article-title>A hyperlink</article-title></element-citation></ref></ref-list></back></article>"
`;

exports[`jats export v1.1: jats-export-1.1 1`] = `
"<!DOCTYPE article PUBLIC \\"-//NLM//DTD JATS (Z39.96) Journal Archiving and Interchange DTD with OASIS Tables with MathML3 v1.1 20151215//EN\\" \\"http://jats.nlm.nih.gov/archiving/1.1/JATS-archive-oasis-article1-mathml3.dtd\\"><article xmlns:xlink=\\"http://www.w3.org/1999/xlink\\"><front><journal-meta><journal-id journal-id-type=\\"publisher-id\\"/><journal-title-group><journal-title/></journal-title-group><issn pub-type=\\"epub\\"/></journal-meta><article-meta><title-group><article-title>Example Manuscript</article-title></title-group><contrib-group content-type=\\"authors\\"><contrib contrib-type=\\"author\\" id=\\"MPContributor_585DB23A-8778-4AFF-986F-CFF7B733CDE6\\"><name><surname>Author</surname><given-names>Example</given-names></name><email>user@example.com</email><xref ref-type=\\"aff\\" rid=\\"MPAffiliation_DED56C09-42E2-4AB4-BA15-9260389E2B08\\"/></contrib><contrib contrib-type=\\"author\\" id=\\"MPContributor_8AD5903E-0C9C-4F46-87AE-124142CE12C2\\" corresp=\\"yes\\"><name><surname>Author</surname><given-names>Corresponding</given-names></name><email>author@example.com</email></contrib></contrib-group><aff id=\\"MPAffiliation_DED56C09-42E2-4AB4-BA15-9260389E2B08\\"><institution>University of Examples</institution><country>United Kingdom</country></aff></article-meta></front><body><sec id=\\"MPSection_DFCBE345-A12D-451E-8320-723F25D2F472\\"><title>Table of Contents</title><div id=\\"MPTOCElement_657C54DB-78B5-4206-E322-170F68263A01\\"/></sec><sec id=\\"MPSection_77786D47-6060-4FBC-BC13-5FA754968D6A\\"><title>First section</title><p id=\\"MPParagraphElement_150780D7-CFED-4529-9398-77B5C7625044\\">The first section.</p><p id=\\"MPParagraphElement_60C13548-23C0-4348-FE98-0510E0B61B18\\">The text in this sentence is <bold>bold</bold>, <italic>italic</italic>, or <underline>underlined</underline>.</p><p id=\\"MPParagraphElement_A89DC62A-90ED-4283-AB6D-06CA2519D801\\"><bold><italic><underline>The text in this sentence is bold, italic and underlined.</underline></italic></bold></p><p id=\\"MPParagraphElement_05A0ED43-8928-4C69-A17C-0A98795001CD\\">The text in this sentence is <sup>superscript</sup> and <sub>subscript</sub>.</p></sec><sec id=\\"MPSection_F8EE617B-484B-4CF2-F1CA-014F5307D5CF\\"><title>Section Two</title><p id=\\"MPParagraphElement_4DF93A63-D44D-41CF-AF54-23712B8703FA\\">The heading of this section is hidden.</p><list list-type=\\"ordered\\"><list-item><p>A numbered list.</p></list-item></list><list list-type=\\"bullet\\"><list-item><p>A bullet list.</p></list-item></list><list list-type=\\"ordered\\"><list-item><p>A numbered list with a nested list.</p><list list-type=\\"ordered\\"><list-item><p>This item is nested.</p></list-item></list></list-item></list><fig-group id=\\"MPFigureElement_A5D68C57-B5BB-4D10-E0C3-ECED717A2AA7\\"><caption><p>A figure with a caption</p></caption><fig id=\\"MPFigure_D673DF08-D75E-4061-8EC1-9611EAB302F0\\"><caption><p/></caption><graphic xlink:href=\\"Data/MPFigure_D673DF08-D75E-4061-8EC1-9611EAB302F0.png\\" mimetype=\\"image\\" mime-subtype=\\"png\\"/></fig></fig-group><fig-group id=\\"MPFigureElement_B32BDEC8-E816-4946-C0BB-0404AF2F0C29\\"><caption><p>A figure with hidden caption</p></caption><fig id=\\"MPFigure_AD65D628-A904-4DC4-A026-F8F4C08F6557\\"><caption><p/></caption><graphic xlink:href=\\"Data/MPFigure_AD65D628-A904-4DC4-A026-F8F4C08F6557.png\\" mimetype=\\"image\\" mime-subtype=\\"png\\"/></fig></fig-group><table-wrap id=\\"MPTableElement_82BD8BAF-D136-4536-AD9C-92C30B9D7315\\"><caption><p>A table</p></caption><table id=\\"MPTable_509A944E-994C-4867-AE55-593628C5DBD5\\"><tr><td>Table Header</td><td>Table Header</td></tr><tr><td>Table Data</td><td>Table Data</td></tr><tr><td>Table Data</td><td>Table Data</td></tr><tr><td>Table Footer</td><td>Table Footer</td></tr></table></table-wrap><table-wrap id=\\"MPTableElement_E0D445B4-18C4-43D3-97B3-1076A7BB5CF4\\"><caption><p>A table with hidden header</p></caption><table id=\\"MPTable_7E83812B-A759-4D38-C1C5-F6507D2FC2A9\\"><tr><td>Table Header</td><td/></tr><tr><td/><td/></tr><tr><td/><td/></tr><tr><td/><td/></tr></table></table-wrap><table-wrap id=\\"MPTableElement_1DF69096-DBA3-47B0-B8C2-E06491E4C655\\"><caption><p>A table with hidden footer</p></caption><table id=\\"MPTable_2A2413E2-71F5-4B6C-F513-7B44748E49A8\\"><tr><td/><td/></tr><tr><td/><td/></tr><tr><td/><td/></tr><tr><td/><td/></tr></table></table-wrap><table-wrap id=\\"MPTableElement_89CF51B2-A20C-4AB9-CBD9-F47DAF167DAA\\"><caption><p>A table with hidden caption</p></caption><table id=\\"MPTable_6C045963-404A-4E1E-FF81-B40E9772A6A6\\"><tr><td/><td/></tr><tr><td/><td/></tr><tr><td/><td/></tr><tr><td/><td/></tr></table></table-wrap><fig id=\\"MPEquationElement_3896A81A-D43B-483F-B990-F64518A940D1\\"><caption><p/></caption><disp-formula><tex-math>2x3=6</tex-math></disp-formula></fig><fig id=\\"MPListingElement_160785A9-4BC0-4C55-C4DE-AF5F4D858DCE\\"><caption><p/></caption><code id=\\"MPListing_4D688DCF-29CD-40CC-93BD-2B13CD883257\\" language=\\"javascript\\">// a JavaScript code block

const x = 'y'</code></fig><p id=\\"MPParagraphElement_345B8B57-2E5E-45ED-E587-4EAAC2BD63E9\\">An inline equation: <inline-formula><tex-math>3+5=8</tex-math></inline-formula></p><p id=\\"MPParagraphElement_6B6DBF43-F7B2-404C-89DD-0EC1E1442B27\\">A cross-reference to a figure: <xref ref-type=\\"fig\\" rid=\\"MPAuxiliaryObjectReference_82E2F9F9-797F-4397-9084-D98B72C12903\\">Figure 1</xref></p><p id=\\"MPParagraphElement_41B99DAB-F86C-477F-D36C-2B8590B5DAD0\\">A cross-reference to a table: <xref ref-type=\\"fig\\" rid=\\"MPAuxiliaryObjectReference_88AF1DCA-B232-4791-8EAA-F38E0025A3C6\\">Table 1</xref></p><p id=\\"MPParagraphElement_C8A173A4-DF32-4242-E800-28A1AE5B48B8\\">A cross-reference to an equation: <xref ref-type=\\"fig\\" rid=\\"MPAuxiliaryObjectReference_461D1AFF-1871-4594-8E07-963D3476B7F5\\">Equation 1</xref></p><p id=\\"MPParagraphElement_CB7347E5-8502-4667-82B9-51104BB1FEDB\\"><xref ref-type=\\"bibr\\" rid=\\"MPBibliographyItem_8C394C86-F7B0-48CE-D5BC-E7A10FCE7FA5\\">A hyperlink</xref></p><p id=\\"MPParagraphElement_3708E8C7-441C-4502-8BFC-D3A2059CEF8B\\">A footnote:<xref ref-type=\\"fn\\" rid=\\"MPFootnote_B1589279-0F3B-472B-BF11-3F59D6B65791\\">1</xref></p><fn-group id=\\"MPFootnotesElement_4CD7CE08-0880-46BD-FF83-52E32AEFAC3E\\"/></sec></body><back><ref-list id=\\"MPSection_E07B0D52-9642-4D58-E577-26F8804E3DEE\\"><title>Bibliography</title><ref id=\\"MPBibliographyItem_8C394C86-F7B0-48CE-D5BC-E7A10FCE7FA5\\"><element-citation publication-type=\\"journal\\"><article-title>A hyperlink</article-title></element-citation></ref></ref-list></back></article>"
`;
